stdlib/README.adoc

```diff
@@ -0,0 +1,102 @@
+= OCaml Stdlib
+
+== Design principles
+
+* When a module `M` has a main type it must be named `M.t`.
+* Functions should take the main type `t` as first argument (unless
+  there is a good reason for it, see next point).
+* If a functions takes another function as argument, that argument
```

Comment:
I highly disagree with this principle. Pointless labelling should be avoided it makes it harder to compose functions.

Reply:
I agree. This is trying to impose JST's principles on the stdlib, which clearly does not follow those principles.

Reply:
The question is more whether something like this is desirable.  @dbuenzli pointed out the difficulty to compose functions.  But there are also advantages: with a label, one can put the functional argument at the end which is easier to read, especially when the function body is more than a couple of lines long.

Reply:
> The question is more whether something like this is desirable. @dbuenzli pointed out the 
difficulty to compose functions.

Labels leak in the type of functions and this prevents many useful FP techniques that are based on lifting functions in other domains. 

> one can put the functional argument at the end which is easier to read, especially when the function body is more than a couple of lines long.

I would rather say this is a disadvantage as it favours the inline style which is not particularly readable. There's a very easy way to solve the problem you mention: if your function doesn't fit on the line never write it inline, give it a name and make a local function. This will actually make it much easier for the reader to decompose your expressions and understand what they are doing rather than having her sift through expressions spanning multiple lines.

Labels make sense a lot of sense in certain contexts, e.g. when a function has many arguments of the same type (typical example `src`, `dst`, or `start`, `stop` for indices, etc.) but again gratuitous use of them especially as `f` for functional arguments is harmful in my opinion.

Reply:
>  if your function doesn't fit on the line never write it inline, give it a name and make a local function.

That's what I personally do but it would be nice to hear others because indeed the choices stdlib make do influence the style of programs one writes.